wxGUI: Add the ProjPicker page to the Location Wizard

HuidaeCho opened the pull request:
This PR adds spatial querying of CRSs to the Location Wizard using the [ProjPicker](https://github.com/HuidaeCho/projpicker).

![image](https://user-images.githubusercontent.com/7456117/127844999-c6ca5d50-d2f6-457e-8536-f1e05313d1fb.png)

![image](https://user-images.githubusercontent.com/7456117/127846032-981056d2-8200-4a67-844e-5212fbb12803.png)

![image](https://user-images.githubusercontent.com/7456117/127844840-50a05462-3f90-435b-9d11-1aaea7b8d3a8.png)

HuidaeCho commented:
Addressed #1253.

petrasovaa commented:
Awesome, thank you! I have yet to actually test it but my initial thoughts are to split the code into gui and python library, so you would have also python/grass/projpicker/ where you would have the code unrelated to gui. If I understood the code correctly, this PR also includes a standalone version, in that case it should be named g.gui.projpicker to be consistent with other standalone gui parts.

Also, we don't need the tk code here.

HuidaeCho commented:
> Awesome, thank you! I have yet to actually test it but my initial thoughts are to split the code into gui and python library, so you would have also python/grass/projpicker/ where you would have the code unrelated to gui.

Split in 37a1c6f.

> If I understood the code correctly, this PR also includes a standalone version, in that case it should be named g.gui.projpicker to be consistent with other standalone gui parts.

That standalone version just happens to be part of `projpicker.py`, but it's not meant to be run from GRASS. Actually, it's not even a GRASS module. I think I'd better move g.projpicker to g.gui.projpicker.
 
> Also, we don't need the tk code here.

Removed in 37a1c6f.

ninsbl commented:
Very nice! Just wondering if users understand that they can identify suitable CRS from a map when they read `Select CRS spatially`?
Maybe `Select CRS in a map` or `Select CRS interactively in a map` is more clear? Just for your consideration. This is a really userfriendly addition in any case...

HuidaeCho commented:
> `Select CRS interactively in a map`

Like this! Thanks for the suggestion.

wenzeslaus commented on python/grass/projpicker/common.py:
```diff
@@ -0,0 +1,126 @@
+    Returns:
+        float or None: Typecasted x in float if successful, None otherwise.
+    """
+    if type(x) != float:
```
I think there is no need to ask. Just try doing the conversion.

```
$ python -m timeit -c "type(10.0) != float"
5000000 loops, best of 5: 82.9 nsec per loop
$ python -m timeit -c "float(10.0)"
5000000 loops, best of 5: 56.9 nsec per loop
```

wenzeslaus commented on python/grass/projpicker/common.py:
```diff
@@ -0,0 +1,126 @@
+    if type(x) != float:
+        try:
+            x = float(x)
+        except:
```
`except ValueError`

HuidaeCho replied:
In this specific case, I intend to catch *any* exceptions.

wenzeslaus replied:
I don't think your intention is to catch SystemExit and KeyboardInterrupt which is what bare except does. PEP 8 recommends against using bare except in its [Programming recommendations section](https://www.python.org/dev/peps/pep-0008/#programming-recommendations). Catching all (normal) exceptions is better expressed as `except Exception`, but the usual recommendation is to use more specific exception or exceptions which documents what you are expecting to happen and what you are not expecting to happen. This is partially because things like StopIteration and SyntaxError [inherit from Exception](https://docs.python.org/3/library/exceptions.html#exception-hierarchy). Another one in the list is NameError, so a wrong variable name (xl instead of x1) in the following code, still leads to the except part to be executed instead of Python reporting that xl does not exist. 

```
x1 = "5"
try:
    float(xl)
except Exception:
    print("All is fine")
```

What are you expecting besides invalid strings (ValueError)? Types which can't be converted to float (TypeError)? An integer (`int`) which is too large to convert ([OverflowError](https://docs.python.org/3/library/functions.html?highlight=float#float))?

HuidaeCho replied:
Makes sense, but I was not able to find the list of errors that `float()` can raise. `help(float)` only mentions `OverflowError` and `ValueError` from its `as_integer_ration()` member function. Curious if `float()` can even throw `SystemExit`, `KeyboardInterrupt`, `StopIteration`, or `SyntaxError` or, in other words, that `try` `except` can catch those when it only encloses a `float()`?

HuidaeCho replied:
For now, catches `ValueError` only.

wenzeslaus commented on python/grass/projpicker/common.py:
```diff
@@ -0,0 +1,126 @@
+    """
+    if type(x) != float:
+        try:
+            x = float(x)
```
return, no need for variable.

wenzeslaus commented on python/grass/projpicker/Makefile:
```diff
@@ -0,0 +1,20 @@
+
+PYFILES := $(patsubst %,$(DSTDIR)/%.py,$(MODULES) __init__)
+PYCFILES := $(patsubst %,$(DSTDIR)/%.pyc,$(MODULES) __init__)
+DSTFILES := $(DSTDIR)/VERSION $(DSTDIR)/projpicker.db
```
Perhaps include a readme with documentation on what this file is and how to update it.

HuidaeCho replied:
README added in 0208123

wenzeslaus commented on gui/wxpython/projpicker_gui/gui_common.py:
```diff
@@ -0,0 +1,250 @@
+
+import grass.projpicker as ppik
+
+projpicker_coordinates_env = "PROJPICKER_COORDINATES"
```
[PEP 8](https://www.python.org/dev/peps/pep-0008/#global-variable-names) recommends to mark globals as private or use `__all__`.

HuidaeCho replied:
Done in 2ca331f

wenzeslaus commented on gui/wxpython/projpicker_gui/wxwidgets.py:
```diff
@@ -0,0 +1,798 @@
+class ProjPickerPanel(wx.Panel):
+    def __init__(self, *args, layout="full", filter_bbox=None, **kwargs):
+        if layout not in ("full", "map_top", "map_left"):
+            raise Exception(f"{layout}: Layout not full, map_top, or map_left")
```
That's a ValueError in Python.

wenzeslaus commented on gui/wxpython/projpicker_gui/wxwidgets.py:
```diff
@@ -0,0 +1,798 @@
+            # query tab
+            self.query_panel = wx.Panel(self.notebook)
+            self.query_panel.page = self.notebook.PageCount
+            self.notebook.AddPage(self.query_panel, "Query")
```
The user-visible text should be translatable: `_("...")`

HuidaeCho replied:
Done in 67c3047

wenzeslaus reviewed (changes requested):
This is great. Thanks. Is this meant for 8.0 or 8.2?

It is a lot of code, so it would be good if it is in best condition possible. Did you try Pylint or similar tools to critique the code?

grass.projpicker subpackage is a good idea. The getosm piece also seems like library material. Basically anything which is not using wxPython is a good candidate. More reusable code in python/grass, great. Less code in GUI, even better.

I just left some mostly minor comments as I was trying to familiarize myself with the code, but the database is something I would like to know more about.

petrasovaa commented:
Seems to work well! When I first opened the wizard it took quite a bit of time, I assume because downloading OSM. Would it be possible to download only when you actually select this option? Or better, do it in separate thread.

The help is little overwhelming with a lot of options. Perhaps we could have a toolbar above the map, similarly to r.li.setup wizard. In any case, I think drawing bbox should be default, not polygon. The search shouldn't be case sensitive.

Overall I feel this should be eventually merged with the second option (Search by EPSG), because there is already an overlap in terms of the functionality and it would make sense from user point of view. That's for a different PR though.

HuidaeCho commented:
> When I first opened the wizard it took quite a bit of time

I think this is because of `self.ppikpage.DoLayout()`?

> I think drawing bbox should be default, not polygon.

I think we just need bbox, no point, no polygon at all. Or maybe keep bbox and point only?

> The search shouldn't be case sensitive.

Search is already case insensitive.

> Overall I feel this should be eventually merged with the second option (Search by EPSG), because there is already an overlap in terms of the functionality and it would make sense from user point of view. That's for a different PR though.

I agreed.

HuidaeCho commented on gui/wxpython/projpicker_gui/gui.py:
```diff
@@ -0,0 +1,752 @@
+import webbrowser
+
+import grass.projpicker as ppik
+if __package__:
```
That `if __package__:` code is for running `projpicker.py` directly from the source directory without installation. I tried the PEP 366 code, but I was not able to run the script from where it is. For example, I installed projpicker `pip install projpicker` and made the above change in `site-packages/projpicker/projpicker.py` like this:
```python
if __name__ == "__main__" and __package__ is None:
    __package__ = "projpicker"

# https://stackoverflow.com/a/49480246/16079666
from .common import (BBox, coor_sep, pos_float_pat, bbox_schema,
                     bbox_columns, is_verbose, get_float)
from . import coor_latlon
from . import coor_xy
```
```bash
projpicker -h    # works

python site-packages/projpicker/projpicker.py -h      # doesn't work
Traceback (most recent call last):
  File "site-packages/projpicker/projpicker.py", line 48, in <module>
    from .common import (BBox, coor_sep, pos_float_pat, bbox_schema,
  File "/home/user/usr/local/lib/python3.7/site-packages/projpicker/projpicker.py", line 48, in <module>
    from .common import (BBox, coor_sep, pos_float_pat, bbox_schema,
ImportError: attempted relative import with no known parent package
```
If we never need to run `projpicker.py` from its installation directory, I think we just need relative imports as long as they are in the same directory or we only need absolute imports because these submodules are always installed together.

petrasovaa commented:
> > When I first opened the wizard it took quite a bit of time
> 
> I think this is because of `self.ppikpage.DoLayout()`?

I guess. I was trying to look what is causing that but I got quickly confused from all the code. There is a lot of gui code, the help text is repeated there 3 times, is that all used at all? The file names could be more helpful (projpicker_gui/gui.py, gui_common.py, wxwidgets.py? ).
> 
> > I think drawing bbox should be default, not polygon.
> 
> I think we just need bbox, no point, no polygon at all. Or maybe keep bbox and point only?

I agree simplicity may be best, bbox could be enough. I am still thinking about the toolbar like in r.li.setup wizard plus adding a tool to clear polygons:

![Selection_182](https://user-images.githubusercontent.com/7494312/128113979-5cea3bdd-cef7-45fe-b5b2-9b8e17ab4686.png)

HuidaeCho commented:
> > > When I first opened the wizard it took quite a bit of time
> > 
> > 
> > I think this is because of `self.ppikpage.DoLayout()`?
> 
> I guess.

I think it is the `DoLayout()` call that starts running the projpicker even before opening the page.
https://github.com/OSGeo/grass/blob/647a5e6b4d2f953546aef3d1d6dc245f0f0c3d45/gui/wxpython/location_wizard/wizard.py#L119

I'm not sure if that's a good idea or necessary to call `Layout()` that early. Downloading is done in a separate thread already.

> I was trying to look what is causing that but I got quickly confused from all the code. There is a lot of gui code, the help text is repeated there 3 times, is that all used at all? The file names could be more helpful (projpicker_gui/gui.py, gui_common.py, wxwidgets.py? ).
> 

Only one copy of the help message is used. We only use the "map_left" layout and these files are still mostly straight copy-paste from the original project. Needs more work. I think I need to turn this PR to a draft to make that clear.

> > > I think drawing bbox should be default, not polygon.
> > 
> > 
> > I think we just need bbox, no point, no polygon at all. Or maybe keep bbox and point only?
> 
> I agree simplicity may be best, bbox could be enough. I am still thinking about the toolbar like in r.li.setup wizard plus adding a tool to clear polygons:
> 
> ![Selection_182](https://user-images.githubusercontent.com/7494312/128113979-5cea3bdd-cef7-45fe-b5b2-9b8e17ab4686.png)

Agreed on bbox only. About the toolbar idea... I need an artist ;-). Right now
* drag for panning
* scroll for zooming
* ctrl+drag for bbox zooming
* click-move-double click for drawing bbox
* click-move-drag-move-double click for drawing bbox with panning
* right click to cancel drawing
* double right click to clear everything

So.. icons, we would need
* zoom in
* zoom out
* pan
* bbox (no more start drawing -> click and pan -> complete; just single click to complete, which is not implemented)
* clear

And remove the Help tab.

HuidaeCho commented:
TODO
* Remove ProjPicker-only code (I'll keep enough code for g.projpicker though, which will include the more full-blown GUI).
* i18nify user-visible messages
* pylint
* black? Other GUI code is not black-formatted yet?

wenzeslaus commented:
> Other GUI code is not black-formatted yet?

Why do you think so? It should be.

HuidaeCho commented:
> > Other GUI code is not black-formatted yet?
> 
> Why do you think so? It should be.

You're right! It was my code in `wizard.py`.

HuidaeCho commented:
> ValueError is likely still the best one here. NotImplementedError has a specific meaning when working with abstract classes/methods, at least that's how it is defined:
> 
> https://docs.python.org/3/library/exceptions.html#NotImplementedError

I think *any* builtin errors are defined specifically for Python internally? For example, `SyntaxError` may not be meant for *custom syntax* like ProjPicker Query. Maybe, it would be better to define my own custom errors (best practice?), but I just prefer not to do that if I can recycle builtin errors (bad idea?).

HuidaeCho commented:
Current shortcuts
```
Pan:               Left drag
Zoom:              Scroll
Zoom to bboxes:    Ctrl + scroll up
Zoom to the world: Ctrl + scroll down
Zoom to a bbox:    Ctrl + left drag
Draw a bbox:       Left click
Cancel drawing:    Right click
Clear:             Double right click
```
No more polygons and points.

ocsmit commented:
@wenzeslaus, @petrasovaa - I have spoken with @HuidaeCho about plans to take over this PR. I am unable to commit directly to the PR at the moment (tried standard git workflow of adding his remote branch, making local branch and then committing to remote branch head), and wanted to inquire what the best approach to start working directly off the PR would be.

wenzeslaus commented:
@ocsmit I don't think there is a easy way of getting access to someones branch except 1) getting access to their repo and 2) when you are a maintainer of the project the PR was submitted to. Conclusion: To overtake the work on this, you need to get this branch locally, push it to your fork and open a new PR.

I would say, no harm done here opening a new PR. There is plenty of comments here, but they seem to be mostly resolved (please check), so opening a new PR will clear the old stuff. Restating and updating the goals and methods in the PR description might be helpful too.
